File: backend/src/model/project.go
package model

import (
	"time"
	"webservices/src/types"
)

type Project struct {
	ID                string                `gorm:"primaryKey;type:uuid;default:uuid_generate_v4()" json:"id" `
	OwnerID           string                `json:"ownerId" `
	Name              string                `gorm:"not null" json:"name" `
	Category          *string               `json:"category,omitempty"`
	Description       *string               `json:"description,omitempty"`
	Image             *string               `json:"image,omitempty"`
	Color             *string               `json:"color,omitempty"`
	Order             int                   `gorm:"column:order_index;default:0" json:"order"`
	Status            types.ProjectStatus   `gorm:"type:project_status;default:'active'" json:"status"`
	LastAssignedIndex *int                  `gorm:"column:last_assigned_index;default:0" json:"lastAssignedIndex,omitempty"`
	CreatedAt         time.Time             `gorm:"column:created_at;default:now();<-:create" json:"createdAt"`
	UpdatedAt         time.Time             `gorm:"column:updated_at;autoUpdateTime" json:"updatedAt"`
	Role              types.UserProjectRole `gorm:"-:all" json:"role,omitempty" comment:"this field obtained from user_projects"`

	Setting     *ProjectSetting  `gorm:"foreignKey:ProjectID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"setting,omitempty"`
	ActiveUsers []User           `gorm:"foreignKey:ProjectID;constraint:OnUpdate:CASCADE,OnDelete:SET NULL;" json:"activeUsers,omitempty"`
	Users       []User           `gorm:"many2many:user_projects;" json:"users,omitempty"`
	Issues      []Issue          `gorm:"constraint:OnDelete:CASCADE;" json:"issues,omitempty"`
	Activities  []RecentActivity `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;" json:"activities,omitempty"`
}

func (Project) TableName() string {
	return "projects"
}

File: frontend/src/types/schemas/project.ts
import { z } from 'zod';
import { User } from '@/types/schemas/user';
import { Setting } from '@/types/schemas/setting';
import { Activity } from '@/types/schemas/activity';
import { USER_ROLES_MAP, UserRoles } from '@/types/misc';

export const projectSchema = z.object({
    name: z.string().min(1, {
        message: "Project name is required"
    }).max(30, 'Maximum 30 characters'),
    image: z.string().optional(),
    color: z.string().optional(),
    description: z.string().optional(),
});

export const projectInviteSchema = z.object({
    email: z.string().min(1, "Email address is required").email("Email address is invalid"),
    role: z.enum(Object.keys(USER_ROLES_MAP) as [UserRoles, ...UserRoles[]]),
    message: z.string().optional(),
})

export type ProjectSchema = z.infer<typeof projectSchema>
export type ProjectInviteSchema = z.infer<typeof projectInviteSchema>
export type ProjectStatus = 'active' | 'inactive' | 'done'

export interface Project extends ProjectSchema {
    id: string
    projectId?: string
    order: number
    status: ProjectStatus
    lastAssignedIndex?: number
    setting?: Setting
    users?: User[]
    activeUsers?: User[]
    issues?: any[]
    createdAt: string
    updatedAt: string
    activities?: Activity[]
    role?: UserRoles
}